File: ROT47/Rust/rot47.rs
#[derive(Default)]
pub(crate) struct ROT47 {
    pub input: String
}

impl ROT47 {
    pub fn new(input: String) -> Self {
        Self { input }
    }

    pub(crate) fn encode(&self) -> String {
        let mut output = String::new();
        for c in self.input.chars() {
            if c.is_ascii() && (c as u8) >= 33 && (c as u8) <= 126 {
                let encoded = ((c as u8 - 33 + 47) % 94) + 33;
                output.push(encoded as char);
            } else {
                output.push(c);
            }
        }
        output
    }

    pub(crate) fn decode(&self, encoded_version: &String) -> String {
        let mut output = String::new();
        for c in encoded_version.chars() {
            if c.is_ascii() && (c as u8) >= 33 && (c as u8) <= 126 {
                let decoded = ((c as u8 - 33 + 47) % 94) + 33;
                output.push(decoded as char);
            } else {
                output.push(c);
            }
        }
        output
    }
}

File: ROT47/C/rot47.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct rot47
{
    size_t len{};
    char *input{};

    void init()
    {
        input = (char*) malloc(1000);
        if (!input)
        {
            exit(1);
        }

        fgets(input, 1000, stdin);
        input[strcspn(input, "\n")] = '\0';
        len = strlen(input);
    }

    char *encode() const
    {
        char *encoded = malloc(len + 1);

        for (size_t i = 0; i < len; i++)
        {
            if (input[i] >= 33 && input[i] <= 126)
            {
                encoded[i] = (char) (33 + (input[i] - 33 + 47) % 94);
            }
        }
        encoded[len] = '\0';
        return encoded;
    }

    char *decode(const char *encoded_version) const
    {
        char *decoded = malloc(len + 1);

        for (size_t i = 0; i < len; i++)
        {
            if (input[i] >= 33 && input[i] <= 126)
            {
                decoded[i] = (char) (33 + (input[i] - 33 + 47) % 94);
            }
        }
        decoded[len] = '\0';
        return decoded;
    }
};
